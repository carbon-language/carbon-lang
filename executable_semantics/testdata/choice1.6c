// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
// Exceptions. See /LICENSE for license information.
// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception

choice Ints {
  None();
  One(Int);
  Two(Int,Int);
}

fn main() -> Int {
  var auto: x = Ints.None();
  var auto: y = Ints.One(42);
  var auto: n = 0;
  match (y) {
    case Ints.None() =>
      n = n + 2;
    case Ints.One(auto: x) =>
      n = x + 1 - 42;
    case Ints.Two(auto: a, auto: b) =>
      n = 2;
  }
  match (x) {
    case Ints.One(auto: x) =>
      n = x + 2;
    case Ints.None() =>
      n = n - 1;
    case Ints.Two(auto: x, auto: y) =>
      n = 5;
  }
  return n;
}
