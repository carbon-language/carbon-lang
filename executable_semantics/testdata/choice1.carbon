// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
// Exceptions. See /LICENSE for license information.
// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception

choice Ints {
  None,
  One(Int),
  Two(Int,Int)
}

fn main() -> Int {
  var x: auto = Ints.None();
  var y: auto = Ints.One(42);
  var n: auto = 0;
  match (y) {
    case Ints.None =>
      n = n + 2;
    case Ints.One(x: auto) =>
      n = x + 1 - 42;
    case Ints.Two(a: auto, b: auto) =>
      n = 2;
  }
  match (x) {
    case Ints.One(x: auto) =>
      n = x + 2;
    case Ints.None() =>
      n = n - 1;
    case Ints.Two(x: auto, y: auto) =>
      n = 5;
  }
  return n;
}

// Test some alternate syntaxes
choice MoreInts {
  None(),
  One(Int),
  Two(Int,Int),
}
