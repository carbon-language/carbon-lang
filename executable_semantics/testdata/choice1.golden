********** source program **********
choice Ints {
alt None ();
alt One Int;
alt Two (0 = Int, 1 = Int);
}
fn main () -> Int {
var auto: x = Ints.None();
var auto: y = Ints.One(0 = 42);
var auto: n = 0;
match (y) {
case Ints.None =>
n = (n + 2);
case Ints.One(0 = auto: x) =>
n = ((x + 1) - 42);
case Ints.Two(0 = auto: a, 1 = auto: b) =>
n = 2;
}
match (x) {
case Ints.One(0 = auto: x) =>
n = (x + 2);
case Ints.None() =>
n = (n - 1);
case Ints.Two(0 = auto: x, 1 = auto: y) =>
n = 5;
}
return n;

}
choice MoreInts {
alt None ();
alt One Int;
alt Two (0 = Int, 1 = Int);
}
********** type checking **********
--- step exp () --->
--- step exp Int --->
--- step exp (0 = Int, 1 = Int) --->
--- step exp Int --->
--- handle value Int with (0 = Int, 1 = Int)<1>(Int,) --->
--- step exp Int --->
--- handle value Int with (0 = Int, 1 = Int)<2>(Int,Int,) --->
--- step exp Int --->
--- step exp () --->
--- step exp Int --->
--- step exp (0 = Int, 1 = Int) --->
--- step exp Int --->
--- handle value Int with (0 = Int, 1 = Int)<1>(Int,) --->
--- step exp Int --->
--- handle value Int with (0 = Int, 1 = Int)<2>(Int,Int,) --->
--- step exp Int --->
--- step exp auto --->
13: type error in call
expected: Int
actual: Tuple(0 = Int)
EXIT CODE: 255
